### ISSUE TITLE
change event with 2 way replication issue + bugfix

### ISSUE THREAD

**Issue by mickael-kerjean** [pouchdb/pouchdb#4627]
Hello there,

First of all, thank you for having created such a cool and usefull project!

As I was having some fun with some code, I went into some issues with the live replication option.
The change event is quite powerfull to plug some application code that can refresh part of screen
depending on the type of document that is received.

With two way sync enable, Everytime some data is push, I receive a push then a pull.
The issue being the docs received while pulling has already been received before.

As a piece of code worth more than just plain text, here is an [example](http://jsfiddle.net/mikeulkeul/0dhft2pa/). The more we insert data, the worst it becomes: see this [other example](http://jsfiddle.net/mikeulkeul/zpc6qjx8/) when trying to insert way more documents.

As it was the perfect opportunity to browse some of pouchdb code, I
drilled down and found where my problem was coming: [/lib/replicate/replicate.js](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js):
1) [Here](https://github.com/pouchdb/pouchdb/blob/8132c76b0a79b75a1ca7f19d1b85943869f6e47d/lib/replicate/replicate.js#L170), we can see how the batch is processed. The interesting functions are [writeDocs](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js#L58) and [finishBatch](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js#L105)
2) the [finishBatch](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js#L105) function trigger the [actuel change event](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js#L110) given a change object that [has the docs field](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js#L109) I was interested in, this field being populated with the changedDocs variable.
3) [writeDocs](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js#L58) is currently resetting the changedDocs array but not [in
every case](https://github.com/pouchdb/pouchdb/blob/master/lib/replicate/replicate.js#L59) which is causing the issue I was experiencing

In other words, for my use case I had to update the original writeDocs
function to become: https://gist.github.com/mickael-kerjean/a6fdda69372ace248b55

That fairly basic but it is now working as expected by my application.

Long life to pouchdb!

**Comment by nolanlawson** [comment]
I'm sorry, but there are a few things about this issue I don't understand:

1. What exactly is the problem? Performance? You don't specify.
2. Can you please provide your fix as a pull request? It's much easier if we can see the diff between the old code and the new code, and plus Travis will automatically run our tests, so we can verify that your change doesn't break anything.

That said, if there are inefficiencies in the live replication code, I'm not surprised at all, and would gladly welcome a fix. :)

**Comment by mickael-kerjean** [comment]
1) Yes it is a performance issue. [This example](http://jsfiddle.net/mikeulkeul/zpc6qjx8/) simulates 2 databases that synchronize together, both are initialize with a set of documents. If we push a batch of documents in one database, the change event is trigger a crazy amount of times (23 times in the example) and the data received from the change event is repeated over and over again. Performance was greatly improve by this small change in the code. 
2) sure, I'll grab a linux machine this week and will write a proper test.

**Comment by daleharvey** [comment]
Fixed in https://github.com/pouchdb/pouchdb/commit/c905239093ed6e7c44b515e851b8cf7987110360, cheers @mickael-kerjean for the detailed report and @willholley for picking it up and fixing it
